torch/nn/parallel/distributed.py

```diff
@@ -570,91 +562,26 @@ def __setstate__(self, state):
-            module_copies[0] = self.module
-
-            for module_copy in module_copies[1:]:
-                for param, copy_param in zip(
-                    self.module.parameters(), self._get_parameters(module_copy)
-                ):
-                    # Reducer requires param copies have the same strides across replicas.
-                    # Fixes up copy_param strides in case replicate didn't match param strides.
-                    if (
-                        param.layout is torch.strided
-                        and param.stride() != copy_param.stride()
-                    ):
-                        with torch.no_grad():
-                            copy_param.set_(
-                                copy_param.clone()
-                                .as_strided(param.size(), param.stride())
-                                .copy_(copy_param)
-                            )
-                    copy_param.requires_grad = param.requires_grad
-
-            return module_copies
-
-        else:
-            return [self.module]
-
     def _build_params_for_reducer(self):
         # Build tuple of (module, parameter) for all parameters that require grads.
-        if self.device_ids and len(self.device_ids) > 1:
-            # Single-process multi-device mode,does not support self.parameters_to_ignore.
-            if self.parameters_to_ignore:
-                raise ValueError(
-                    "Single-Process multi-device mode does not "
-                    "support ignoring parameters upfront. Please consider "
-                    "using one DDP instance per device."
-                )
-
-            modules_and_parameters = [
-                [
-                    (module, parameter)
-                    for module in replica.modules()
-                    for parameter in filter(
-                        lambda parameter: parameter.requires_grad,
-                        self._get_parameters(module, recurse=False),
-                    )
-                ]
-                for replica in self._module_copies
-            ]
-        else:
-            modules_and_parameters = [
-                [
-                    (module, parameter)
-                    for module_name, module in replica.named_modules()
-                    for parameter in [
-                        param
-                        # Note that we access module.named_parameters instead of
-                        # parameters(module). parameters(module) is only needed in the
-                        # single-process multi device case, where it accesses replicated
-                        # parameters through _former_parameters.
-                        for param_name, param in module.named_parameters(recurse=False)
-                        if param.requires_grad
-                        and f"{module_name}.{param_name}"
-                        not in self.parameters_to_ignore
-                    ]
+        modules_and_parameters = [
```

Comment:
This outermost list is always one element now? I guess to flatten in we'd need to change the reducer signature though?

Reply:
As the description of this PR says, I plan to deprecate` _module_copies` field, which is the root cause of the extra complexity here. Will simplify it in the next PR.